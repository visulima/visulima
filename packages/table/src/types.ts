export type HorizontalAlign = "center" | "left" | "right";
export type VerticalAlign = "bottom" | "middle" | "top";

export type Border = {
    body: string;
    left: string;
    middle: string;
    right: string;
};

export type CellOptions = {
    colSpan?: number;
    content: number | string | null | undefined;
    hAlign?: HorizontalAlign;
    rowSpan?: number;
    vAlign?: VerticalAlign;
};

export type Cell = CellOptions | number | string | null | undefined;

export type BorderStyle = {
    bodyJoin?: string;
    bodyLeft?: string;
    bodyRight?: string;
    bottomBody?: string;

    bottomJoin?: string;
    bottomLeft?: string;
    bottomRight?: string;
    joinBody?: string;

    joinJoin?: string;
    joinLeft?: string;
    joinRight?: string;

    topBody?: string;
    topJoin?: string;
    topLeft?: string;
    topRight?: string;
};

export type TableOptions = {
    /**
     * The default alignment of the cell content
     */
    align?: HorizontalAlign;
    /**
     * The style of the table borders
     */
    border?: BorderStyle;
    /**
     * Whether to draw the outer border of the table
     */
    drawOuterBorder?: boolean;
    /**
     * The character to use for empty cells
     */
    emptyCellChar?: string;
    /**
     * The maximum width of a cell before truncation
     */
    maxWidth?: number;
    /**
     * The padding between the cell content and the cell border
     */
    padding?: number;
    /**
     * Whether to show the header of the table
     */
    showHeader?: boolean;
    /**
     * Style options for specific parts of the table
     */
    style?: {
        cells?: string[];
        header?: string[];
    };
    /**
     * Whether to truncate cells that are too long
     */
    truncate?: boolean;
    /**
     * The default vertical alignment of the cell content
     */
    vAlign?: VerticalAlign;
};
