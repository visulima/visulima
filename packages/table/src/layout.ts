import type { Cell as CellType, LayoutCell, TableLayout } from "./types";

/** Creates a layout cell from a given cell configuration. */
const createLayoutCell = (cell: CellType, column: number, row: number): LayoutCell => {
    const normalizedCell = typeof cell === "object" && cell !== null ? cell : { content: String(cell) };

    return {
        ...normalizedCell,
        content: String(normalizedCell.content ?? ""),
        height: normalizedCell.rowSpan ?? 1,
        width: normalizedCell.colSpan ?? 1,
        x: column,
        y: row,
    };
}

/**
 * Creates a list of layout cells from a 2D array of rows (CellType[][]).
 * Also inserts "span cells" (placeholders) for the covered columns/rows.
 */
// eslint-disable-next-line sonarjs/cognitive-complexity
const createTableLayout = (rows: CellType[][]): TableLayout => {
    // Step 1: figure out the total columns by max sum of colSpans in any row
    let maxCols = 0;
    for (const row of rows) {
        let sum = 0;
        for (const cell of row) {
            if (cell === null || cell === undefined) {
                sum += 1;
            } else if (typeof cell === "object" && !Array.isArray(cell)) {
                sum += cell.colSpan ?? 1;
            } else {
                sum += 1;
            }
        }
        maxCols = Math.max(maxCols, sum);
    }

    // Step 2: build out the cells
    const layoutCells: LayoutCell[] = [];
    let rowIndex = 0;
    
    // Create a grid to track occupied positions
    const occupiedPositions: Record<string, LayoutCell> = {};
    
    // Helper function to check if a position is occupied
    const isPositionOccupied = (x: number, y: number): boolean => {
        return occupiedPositions[`${x},${y}`] !== undefined;
    };
    
    // Helper function to mark a position as occupied
    const markPositionOccupied = (x: number, y: number, cell: LayoutCell): void => {
        occupiedPositions[`${x},${y}`] = cell;
    };

    for (const row of rows) {
        // We track where we place each cell. colPointer moves left to right.
        let colPointer = 0;
        for (const cellValue of row) {
            if (cellValue == null) {
                // This 'slot' is presumably covered by a spanning cell
                colPointer += 1;
                continue;
            }
            
            // Skip over any positions already occupied by spans from earlier rows
            while (colPointer < maxCols && isPositionOccupied(colPointer, rowIndex)) {
                colPointer += 1;
            }
            
            // If we reached the end of the row, break out
            if (colPointer >= maxCols) {
                break;
            }
            
            // Create a layout cell for this position
            const layoutCell = createLayoutCell(cellValue, colPointer, rowIndex);
            layoutCells.push(layoutCell);
            
            // Mark all positions covered by this cell as occupied
            for (let ry = rowIndex; ry < rowIndex + layoutCell.height; ry++) {
                for (let rx = colPointer; rx < colPointer + layoutCell.width; rx++) {
                    markPositionOccupied(rx, ry, layoutCell);
                }
            }

            // Insert placeholder cells for all covered positions except the top-left
            for (let ry = rowIndex; ry < rowIndex + layoutCell.height; ry++) {
                for (let rx = colPointer; rx < colPointer + layoutCell.width; rx++) {
                    if (rx === colPointer && ry === rowIndex) {
                        // the real cell - already added
                        continue;
                    }
                    // placeholder cell
                    layoutCells.push({
                        content: "",
                        height: 1,
                        isSpanCell: true,
                        parentCell: layoutCell,
                        width: 1,
                        x: rx,
                        y: ry,
                    });
                }
            }

            colPointer += layoutCell.width;
        }
        rowIndex++;
    }

    // Step 3: total table height is the number of rows
    // (But some cells might extend downward. We'll find the max.)
    let maxRow = rows.length;
    for (const c of layoutCells) {
        maxRow = Math.max(maxRow, c.y + c.height);
    }

    // Step 4: return the structure
    return {
        cells: layoutCells,
        height: maxRow,
        width: maxCols,
    };
}

export default createTableLayout;
