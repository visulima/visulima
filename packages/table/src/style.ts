import type { BorderStyle } from "./types";

/**
 * Default border style using simple ASCII characters.
 */
export const DEFAULT_BORDER: BorderStyle = {
    bodyJoin: "│",
    bodyLeft: "│",
    bodyRight: "│",
    bottomBody: "─",
    bottomJoin: "┴",
    bottomLeft: "└",
    bottomRight: "┘",
    joinBody: "─",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "─",
    topJoin: "┬",
    topLeft: "┌",
    topRight: "┐",
};

/**
 * Minimal border style using only horizontal lines.
 */
export const MINIMAL_BORDER: BorderStyle = {
    bodyJoin: "",
    bodyLeft: "",
    bodyRight: "",
    bottomBody: "-",
    bottomJoin: "-",
    bottomLeft: "-",
    bottomRight: "-",
    joinBody: "-",
    joinJoin: "-",
    joinLeft: "-",
    joinRight: "-",
    topBody: "-",
    topJoin: "-",
    topLeft: "-",
    topRight: "-",
};

/**
 * Double-line border style using Unicode box-drawing characters.
 */
export const DOUBLE_BORDER: BorderStyle = {
    bodyJoin: "║",
    bodyLeft: "║",
    bodyRight: "║",
    bottomBody: "═",
    bottomJoin: "╩",
    bottomLeft: "╚",
    bottomRight: "╝",
    joinBody: "═",
    joinJoin: "╬",
    joinLeft: "╠",
    joinRight: "╣",
    topBody: "═",
    topJoin: "╦",
    topLeft: "╔",
    topRight: "╗",
};

/**
 * Border style with rounded corners using Unicode box-drawing characters.
 */
export const ROUNDED_BORDER: BorderStyle = {
    bodyJoin: "│",
    bodyLeft: "│",
    bodyRight: "│",
    bottomBody: "─",
    bottomJoin: "┴",
    bottomLeft: "╰",
    bottomRight: "╯",
    joinBody: "─",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "─",
    topJoin: "┬",
    topLeft: "╭",
    topRight: "╮",
};

/**
 * Border style using dots for the border.
 */
export const DOTS_BORDER: BorderStyle = {
    bodyJoin: "┊",
    bodyLeft: "┊",
    bodyRight: "┊",
    bottomBody: "┈",
    bottomJoin: "┴",
    bottomLeft: "└",
    bottomRight: "┘",
    joinBody: "┈",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "┈",
    topJoin: "┬",
    topLeft: "┌",
    topRight: "┐",
};

/**
 * Border style using Markdown syntax.
 */
export const MARKDOWN_BORDER: BorderStyle = {
    bodyJoin: "|",
    bodyLeft: "|",
    bodyRight: "|",
    bottomBody: "-",
    bottomJoin: "|",
    bottomLeft: "|",
    bottomRight: "|",
    joinBody: "-",
    joinJoin: "|",
    joinLeft: "|",
    joinRight: "|",
    topBody: "-",
    topJoin: "|",
    topLeft: "|",
    topRight: "|",
};

/**
 * ASCII border style using only ASCII characters.
 */
export const ASCII_BORDER: BorderStyle = {
    bodyJoin: "|",
    bodyLeft: "|",
    bodyRight: "|",
    bottomBody: "-",
    bottomJoin: "+",
    bottomLeft: "+",
    bottomRight: "+",
    joinBody: "-",
    joinJoin: "+",
    joinLeft: "+",
    joinRight: "+",
    topBody: "-",
    topJoin: "+",
    topLeft: "+",
    topRight: "+",
};

/**
 * No border style.
 */
export const NO_BORDER: BorderStyle = {
    bodyJoin: "",
    bodyLeft: "",
    bodyRight: "",
    bottomBody: "",
    bottomJoin: "",
    bottomLeft: "",
    bottomRight: "",
    joinBody: "",
    joinJoin: "",
    joinLeft: "",
    joinRight: "",
    topBody: "",
    topJoin: "",
    topLeft: "",
    topRight: "",
};
