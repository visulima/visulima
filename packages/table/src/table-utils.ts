import type { Cell as CellType , LayoutCell } from "./types";

/** Computes the logical width of a row (the sum of colSpans, defaulting to 1 per cell). */
export function computeRowLogicalWidth(row: CellType[]): number {
    // eslint-disable-next-line unicorn/no-array-reduce
    return row.reduce((total: number, cell: CellType) => {
        if (cell === null) {
            return total + 1;
        }

        if (typeof cell === "object" && !Array.isArray(cell)) {
            return total + (cell.colSpan ?? 1);
        }

        return total + 1;
    }, 0);
}

/** Pads a row with null cells so that its logical width equals targetWidth. */
export function fillRowToWidth(row: CellType[], targetWidth: number): CellType[] {
    const currentWidth = computeRowLogicalWidth(row);

    if (currentWidth < targetWidth) {
        return [...row, ...Array.from({length: targetWidth - currentWidth}).fill(null)] as CellType[];
    }

    return row;
}

/** Returns the underlying "real" cell (if a span cell, returns its parent). */
export function getRealCell(layoutCell: LayoutCell | null): LayoutCell | null {
    if (!layoutCell) {
        return null;
    }

    if (layoutCell.isSpanCell && layoutCell.parentCell) {
        return layoutCell.parentCell;
    }

    return layoutCell;
}

/** Checks whether two layout cells represent the same originating cell. */
export function areCellsEquivalent(cellA: LayoutCell | null, cellB: LayoutCell | null): boolean {
    const realA = getRealCell(cellA);
    const realB = getRealCell(cellB);
    return realA === realB;
}
