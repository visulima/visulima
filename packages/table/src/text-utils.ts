import { stripVTControlCharacters } from "node:util";
import ansiRegex from "ansi-regex";
import stringWidth from "string-width";
import type { TruncateOptions } from "./types";

const globalAnsiPattern = ansiRegex();

/** Finds the real (nonâ€‘ANSI) character index in text corresponding to the visible position. */
function findRealPosition(text: string, visiblePosition: number): number {
    let visibleIndex = 0;
    let match: RegExpExecArray | null = null;
    const ansiRanges: { end: number; start: number }[] = [];
    globalAnsiPattern.lastIndex = 0;
    while ((match = globalAnsiPattern.exec(text)) !== null) {
        ansiRanges.push({ end: match.index + match[0].length, start: match.index });
    }
    let currentIndex = 0;
    while (currentIndex < text.length) {
        const range = ansiRanges.find((r) => currentIndex >= r.start && currentIndex < r.end);
        if (range) {
            currentIndex = range.end;
            continue;
        }
        const charWidth = stringWidth(text[currentIndex]);
        if (visibleIndex + charWidth > visiblePosition) {
            return currentIndex;
        }
        visibleIndex += charWidth;
        currentIndex++;
    }
    return Math.min(stringWidth(text), visiblePosition) - 1;
}

/** Finds the index of the nearest space character to a target position. */
function getIndexOfNearestSpace(text: string, targetIndex: number, searchRight = false): number {
    if (text.charAt(targetIndex) === " ") {
        return targetIndex;
    }
    const direction = searchRight ? 1 : -1;
    for (let offset = 0; offset <= 3; offset++) {
        const pos = targetIndex + offset * direction;
        if (text.charAt(pos) === " ") {
            return pos;
        }
    }
    return targetIndex;
}

/** Preserves ANSI codes in a text slice. */
function preserveAnsiCodes(text: string, startIndex: number, endIndex: number): string {
    const openCodes: string[] = [];
    let match: RegExpExecArray | null = null;
    globalAnsiPattern.lastIndex = 0;
    while ((match = globalAnsiPattern.exec(text)) !== null) {
        if (match.index > endIndex) {
            break;
        }
        const code = match[0];
        if (code === "\u001B[0m") {
            if (match.index < endIndex) {
                openCodes.length = 0;
            }
        } else if (code.startsWith("\u001B[") && match.index < endIndex) {
            openCodes.push(code);
        }
    }
    const slicedText = text.slice(startIndex, endIndex);
    return openCodes.join("") + slicedText + (openCodes.length > 0 ? "\u001B[0m" : "");
}

/** Truncates text to a specified width while preserving ANSI codes. */
export function truncateText(text: string, maxWidth: number, options: Required<TruncateOptions>): string {
    if (typeof text !== "string") {
        throw new TypeError(`Expected input to be a string, got ${typeof text}`);
    }
    if (typeof maxWidth !== "number") {
        throw new TypeError(`Expected maxWidth to be a number, got ${typeof maxWidth}`);
    }
    if (maxWidth < 1) {
        return "";
    }
    if (maxWidth === 1) {
        return options.truncationCharacter;
    }
    const visibleLength = stringWidth(text);
    if (visibleLength <= maxWidth) {
        return text;
    }
    const lines = text.split("\n");
    let { truncationCharacter } = options;
    const truncatedLines = lines.map((line) => {
        const lineLength = stringWidth(line);
        if (lineLength <= maxWidth) {
            return line;
        }
        if (options.position === "start") {
            if (options.preferTruncationOnSpace) {
                const nearestSpace = getIndexOfNearestSpace(line, lineLength - maxWidth + 1, true);
                return truncationCharacter + preserveAnsiCodes(line, nearestSpace, line.length).trim();
            }
            if (options.space) {
                truncationCharacter += " ";
            }
            const visibleStart = stringWidth(line) - maxWidth + stringWidth(truncationCharacter);
            const realStart = findRealPosition(line, visibleStart);
            return truncationCharacter + preserveAnsiCodes(line, realStart, line.length);
        }
        if (options.position === "middle") {
            if (options.space) {
                truncationCharacter = ` ${truncationCharacter} `;
            }
            const halfWidth = Math.floor(maxWidth / 2);
            if (options.preferTruncationOnSpace) {
                const spaceNearFirst = getIndexOfNearestSpace(line, halfWidth);
                const spaceNearSecond = getIndexOfNearestSpace(line, line.length - (maxWidth - halfWidth) + 1, true);
                return (
                    preserveAnsiCodes(line, 0, spaceNearFirst) +
                    truncationCharacter +
                    preserveAnsiCodes(line, spaceNearSecond, line.length).trim()
                );
            }
            const firstHalf = findRealPosition(line, halfWidth);
            const secondHalfStart = stringWidth(line) - (maxWidth - halfWidth) + stringWidth(truncationCharacter);
            const secondHalf = findRealPosition(line, secondHalfStart);
            return preserveAnsiCodes(line, 0, firstHalf) + truncationCharacter + preserveAnsiCodes(line, secondHalf, line.length);
        }
        if (options.position === "end") {
            if (options.preferTruncationOnSpace) {
                const nearestSpace = getIndexOfNearestSpace(line, maxWidth - 1);
                return preserveAnsiCodes(line, 0, nearestSpace) + truncationCharacter;
            }
            if (options.space) {
                truncationCharacter = ` ${truncationCharacter}`;
            }
            const realEnd = findRealPosition(line, maxWidth - stringWidth(truncationCharacter));
            return preserveAnsiCodes(line, 0, realEnd) + truncationCharacter;
        }
        throw new Error(`Expected options.position to be either 'start', 'middle' or 'end', got ${options.position}`);
    });
    return truncatedLines.join("\n");
}

/** Word wraps text to a specified width while preserving ANSI codes. */
export function wordWrapText(text: string, maxWidth: number): string[] {
    if (maxWidth <= 0 || !text) {
        return [text];
    }
    if (stringWidth(text) <= maxWidth) {
        return [text];
    }
    const lines = text.split(/\r?\n/);
    const wrappedLines: string[] = [];
    const colorPattern = ansiRegex();
    const linkPattern = /\u001B\]8;;([^\u0007]*)\u0007([^\u0007]*)\u001B\]8;;\u0007/g;
    for (const line of lines) {
        if (!line.trim()) {
            wrappedLines.push(line);
            continue;
        }
        const formats: { index: number; sequence: string }[] = [];
        let plainText = line;
        let match: RegExpExecArray | null = null;
        colorPattern.lastIndex = 0;
        while ((match = colorPattern.exec(line)) !== null) {
            formats.push({ index: match.index, sequence: match[0] });
        }
        linkPattern.lastIndex = 0;
        while ((match = linkPattern.exec(line)) !== null) {
            formats.push({
                index: match.index,
                sequence: `\u001B]8;;${match[1]}\u0007${match[2]}\u001B\]8;;\u0007`,
            });
        }
        formats.sort((a, b) => a.index - b.index);
        plainText = stripVTControlCharacters(line);
        const words = plainText.split(/\s+/);
        let currentLine = "";
        let currentLineWidth = 0;
        let lastAnsi = "";
        for (const word of words) {
            const wordWidth = stringWidth(word);
            if (currentLineWidth + wordWidth + (currentLine ? 1 : 0) > maxWidth && currentLine) {
                if (lastAnsi) {
                    currentLine += "\u001B[0m";
                }
                wrappedLines.push(currentLine);
                currentLine = "";
                currentLineWidth = 0;
            }
            if (currentLine) {
                currentLine += " ";
                currentLineWidth += 1;
            }
            if (lastAnsi) {
                currentLine += lastAnsi;
            }
            const wordStart = plainText.indexOf(word);
            let formattedWord = word;
            for (const format of formats) {
                if (format.index <= wordStart) {
                    if (format.sequence.startsWith("\u001B[")) {
                        lastAnsi = format.sequence;
                    }
                    formattedWord = format.sequence + formattedWord;
                }
            }
            currentLine += formattedWord;
            currentLineWidth += wordWidth;
        }
        if (currentLine) {
            if (lastAnsi) {
                currentLine += "\u001B[0m";
            }
            wrappedLines.push(currentLine);
        }
    }
    return wrappedLines;
}
