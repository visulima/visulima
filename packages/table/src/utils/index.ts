export { areCellsEquivalent } from "./are-cells-equivalent";
export { computeRowLogicalWidth } from "./compute-row-logical-width";
export { fillRowToWidth } from "./fill-row-to-width";
export { findRealPosition } from "./find-real-position";
export { getIndexOfNearestSpace } from "./get-index-of-nearest-space";
export { getRealCell } from "./get-real-cell";
export { preserveAnsiCodes } from "./preserve-ansi-codes";
export { truncateText } from "./truncate-text";
export { wordWrapText } from "./word-wrap-text";
