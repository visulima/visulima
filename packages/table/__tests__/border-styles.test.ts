import { describe, expect, it } from "vitest";

import { createTable } from "../src";
import { ASCII_BORDER, DEFAULT_BORDER, DOTS_BORDER, DOUBLE_BORDER, MARKDOWN_BORDER, MINIMAL_BORDER, NO_BORDER, ROUNDED_BORDER } from "../src/style";

describe("table Border Styles", () => {
    const sampleData = [
        ["Header 1", "Header 2", "Header 3"],
        ["Row 1 Cell 1", "Row 1 Cell 2", "Row 1 Cell 3"],
        ["Row 2 Cell 1", "Row 2 Cell 2", "Row 2 Cell 3"],
    ];

    // Use it.each to test all border styles
    it.each([
        ["DEFAULT", DEFAULT_BORDER],
        ["MINIMAL", MINIMAL_BORDER],
        ["DOUBLE", DOUBLE_BORDER],
        ["ROUNDED", ROUNDED_BORDER],
        ["DOTS", DOTS_BORDER],
        ["MARKDOWN", MARKDOWN_BORDER],
        ["ASCII", ASCII_BORDER],
        ["NO_BORDER", NO_BORDER],
    ])('renders table with %s border style', (name, borderStyle) => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: borderStyle,
            },
        });

        table.setHeaders(sampleData[0]);
        table.addRows(...sampleData.slice(1));

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    // Test complex scenarios
    it("renders table with spanning cells using DEFAULT border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: DEFAULT_BORDER,
            },
        });

        table.setHeaders(["Spanning Header", "Regular Header"]);
        table.addRows([{ colSpan: 2, content: "Spanning Content" }], ["Regular Cell 1", "Regular Cell 2"]);

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with multiline content using DOUBLE border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: DOUBLE_BORDER,
            },
        });

        table.setHeaders(["Multiline Header", "Regular Header"]);
        table.addRows(["Line 1\\nLine 2\\nLine 3", "Regular Cell"], ["Regular Cell 1", "Regular Cell 2"]);

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with mixed alignments using ROUNDED border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: ROUNDED_BORDER,
            },
        });

        table.setHeaders([
            { content: "Left Align", hAlign: "left" },
            { content: "Center Align", hAlign: "center" },
            { content: "Right Align", hAlign: "right" },
        ]);
        table.addRows(["Short", "Medium Text", "Longer Text Here"], ["A", "BB", "CCC"]);

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with vertical alignments using DOTS border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: DOTS_BORDER,
            },
        });

        table.setHeaders([
            { content: "Top\\nAlign", vAlign: "top" },
            { content: "Middle\\nAlign", vAlign: "middle" },
            { content: "Bottom\\nAlign", vAlign: "bottom" },
        ]);
        table.addRows([
            { content: "Short", vAlign: "top" },
            { content: "Medium\\nText", vAlign: "middle" },
            { content: "Long\\nText\\nHere", vAlign: "bottom" },
        ]);

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with word wrapping using MARKDOWN border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: MARKDOWN_BORDER,
            },
            wordWrap: true,
        });

        table.setHeaders(["Short Header", "Long Header That Should Wrap"]);
        table.addRows(
            ["Short", "This is a very long cell content that should be wrapped across multiple lines"],
            ["Also Short", "Another long content that needs to be wrapped to maintain readability"],
        );

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with truncated content using ASCII border style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: ASCII_BORDER,
            },
        });

        table.setHeaders([{ content: "Truncated Header", maxWidth: 10 }, { content: "Regular Header" }]);
        table.addRows(
            [{ content: "This content should be truncated", maxWidth: 10 }, "Regular Cell"],
            [{ content: "Also truncated content here", maxWidth: 10 }, "Another Regular Cell"],
        );

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });

    it("renders table with mixed styling using NO_BORDER style", () => {
        expect.assertions(1);
        const table = createTable({
            style: {
                border: NO_BORDER,
                paddingLeft: 2,
                paddingRight: 2,
            },
        });

        table.setHeaders([
            { content: "Mixed", hAlign: "left", vAlign: "top" },
            { content: "Styling", hAlign: "center", vAlign: "middle" },
            { content: "Example", hAlign: "right", vAlign: "bottom" },
        ]);
        table.addRows([
            { content: "Left\\nTop", hAlign: "left", vAlign: "top" },
            { content: "Center\\nMiddle", hAlign: "center", vAlign: "middle" },
            { content: "Right\\nBottom", hAlign: "right", vAlign: "bottom" },
        ]);

        const output = table.toString();
        expect(output).toMatchSnapshot();
    });
});
