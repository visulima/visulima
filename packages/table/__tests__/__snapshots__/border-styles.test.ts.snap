// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`table Border Styles > renders table with ASCII border style 1`] = `
"+--------------+--------------+--------------+
| Header 1     | Header 2     | Header 3     |
+--------------+--------------+--------------+
| Row 1 Cell 1 | Row 1 Cell 2 | Row 1 Cell 3 |
+--------------+--------------+--------------+
| Row 2 Cell 1 | Row 2 Cell 2 | Row 2 Cell 3 |
+--------------+--------------+--------------+"
`;

exports[`table Border Styles > renders table with DEFAULT border style 1`] = `
"┌──────────────┬──────────────┬──────────────┐
│ Header 1     │ Header 2     │ Header 3     │
├──────────────┼──────────────┼──────────────┤
│ Row 1 Cell 1 │ Row 1 Cell 2 │ Row 1 Cell 3 │
├──────────────┼──────────────┼──────────────┤
│ Row 2 Cell 1 │ Row 2 Cell 2 │ Row 2 Cell 3 │
└──────────────┴──────────────┴──────────────┘"
`;

exports[`table Border Styles > renders table with DOTS border style 1`] = `
"┌┈┈┈┈┈┈┈┈┈┈┈┈┈┈┬┈┈┈┈┈┈┈┈┈┈┈┈┈┈┬┈┈┈┈┈┈┈┈┈┈┈┈┈┈┐
┊ Header 1     ┊ Header 2     ┊ Header 3     ┊
├┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┤
┊ Row 1 Cell 1 ┊ Row 1 Cell 2 ┊ Row 1 Cell 3 ┊
├┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┤
┊ Row 2 Cell 1 ┊ Row 2 Cell 2 ┊ Row 2 Cell 3 ┊
└┈┈┈┈┈┈┈┈┈┈┈┈┈┈┴┈┈┈┈┈┈┈┈┈┈┈┈┈┈┴┈┈┈┈┈┈┈┈┈┈┈┈┈┈┘"
`;

exports[`table Border Styles > renders table with DOUBLE border style 1`] = `
"╔══════════════╦══════════════╦══════════════╗
║ Header 1     ║ Header 2     ║ Header 3     ║
╠══════════════╬══════════════╬══════════════╣
║ Row 1 Cell 1 ║ Row 1 Cell 2 ║ Row 1 Cell 3 ║
╠══════════════╬══════════════╬══════════════╣
║ Row 2 Cell 1 ║ Row 2 Cell 2 ║ Row 2 Cell 3 ║
╚══════════════╩══════════════╩══════════════╝"
`;

exports[`table Border Styles > renders table with MARKDOWN border style 1`] = `
"|--------------|--------------|--------------|
| Header 1     | Header 2     | Header 3     |
|--------------|--------------|--------------|
| Row 1 Cell 1 | Row 1 Cell 2 | Row 1 Cell 3 |
|--------------|--------------|--------------|
| Row 2 Cell 1 | Row 2 Cell 2 | Row 2 Cell 3 |
|--------------|--------------|--------------|"
`;

exports[`table Border Styles > renders table with MINIMAL border style 1`] = `
"----------------------------------------------
 Header 1      Header 2      Header 3     
----------------------------------------------
 Row 1 Cell 1  Row 1 Cell 2  Row 1 Cell 3 
----------------------------------------------
 Row 2 Cell 1  Row 2 Cell 2  Row 2 Cell 3 
----------------------------------------------"
`;

exports[`table Border Styles > renders table with NO_BORDER border style 1`] = `
" Header 1      Header 2      Header 3     
 Row 1 Cell 1  Row 1 Cell 2  Row 1 Cell 3 
 Row 2 Cell 1  Row 2 Cell 2  Row 2 Cell 3 "
`;

exports[`table Border Styles > renders table with ROUNDED border style 1`] = `
"╭──────────────┬──────────────┬──────────────╮
│ Header 1     │ Header 2     │ Header 3     │
├──────────────┼──────────────┼──────────────┤
│ Row 1 Cell 1 │ Row 1 Cell 2 │ Row 1 Cell 3 │
├──────────────┼──────────────┼──────────────┤
│ Row 2 Cell 1 │ Row 2 Cell 2 │ Row 2 Cell 3 │
╰──────────────┴──────────────┴──────────────╯"
`;

exports[`table Border Styles > renders table with mixed alignments using ROUNDED border style 1`] = `
"╭────────────┬──────────────┬──────────────────╮
│ Left Align │ Center Align │      Right Align │
├────────────┼──────────────┼──────────────────┤
│ Short      │ Medium Text  │ Longer Text Here │
├────────────┼──────────────┼──────────────────┤
│ A          │ BB           │ CCC              │
╰────────────┴──────────────┴──────────────────╯"
`;

exports[`table Border Styles > renders table with mixed styling using NO_BORDER style 1`] = `
"  Mixed        Styling           Example        
  Left\\nTop    Center\\nMiddle    Right\\nBottom  "
`;

exports[`table Border Styles > renders table with multiline content using DOUBLE border style 1`] = `
"╔════════════════════════╦════════════════╗
║ Multiline Header       ║ Regular Header ║
╠════════════════════════╬════════════════╣
║ Line 1\\nLine 2\\nLine 3 ║ Regular Cell   ║
╠════════════════════════╬════════════════╣
║ Regular Cell 1         ║ Regular Cell 2 ║
╚════════════════════════╩════════════════╝"
`;

exports[`table Border Styles > renders table with spanning cells using DEFAULT border style 1`] = `
"┌─────────────────┬────────────────┐
│ Spanning Header │ Regular Header │
├──────────────────────────────────┤
│ Spanning Content                 │
├─────────────────┼────────────────┤
│ Regular Cell 1  │ Regular Cell 2 │
└─────────────────┴────────────────┘"
`;

exports[`table Border Styles > renders table with truncated content using ASCII border style 1`] = `
"+------------+----------------------+
| Truncated… | Regular Header       |
+------------+----------------------+
| This cont… | Regular Cell         |
+------------+----------------------+
| Also trun… | Another Regular Cell |
+------------+----------------------+"
`;

exports[`table Border Styles > renders table with vertical alignments using DOTS border style 1`] = `
"┌┈┈┈┈┈┈┈┈┈┈┈┈┬┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┬┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┐
┊ Top\\nAlign ┊ Middle\\nAlign ┊ Bottom\\nAlign    ┊
├┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┤
┊ Short      ┊ Medium\\nText  ┊ Long\\nText\\nHere ┊
└┈┈┈┈┈┈┈┈┈┈┈┈┴┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┴┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┘"
`;

exports[`table Border Styles > renders table with word wrapping using MARKDOWN border style 1`] = `
"|--------|-------------|
| Short  | Long Header |
| Header | That Should |
|        | Wrap        |
|--------|-------------|
| Short  | This is a   |
|        | very long   |
|        | cell        |
|        | content     |
|        | that should |
|        | be wrapped  |
|        | across      |
|        | multiple    |
|        | lines       |
|--------|-------------|
| Also   | Another     |
| Short  | long        |
|        | content     |
|        | that needs  |
|        | to be       |
|        | wrapped to  |
|        | maintain    |
|        | readability |
|--------|-------------|"
`;
